File: Striver A2Z/Step 7 - Recursion [PatternWise]/Lec 2 - Subsequences Pattern/generate-parentheses.cpp

// https://leetcode.com/problems/generate-parentheses/

class Solution {
public:

    vector<string> ans;

    void solve(int n, int open, int close, string curr){
        if(open == n && close == n){
            ans.push_back(curr);
            return;
        }

        if(open < n) solve(n, open+1, close, curr+'(');
        if(open > close) solve(n, open, close+1, curr+')');
        return;
    }

    vector<string> generateParenthesis(int n) {
        // vector<string> ans;
        // if(!n) return ans;
        int open = 0, close = 0;
        solve(n, open, close, "");
        return ans;
    }
};

File: Neetcode-150-js/Stack/generate-parentheses.js

// https://leetcode.com/problems/generate-parentheses/description/

/**
 * @param {number} n
 * @return {string[]}
 */

var helper = function(n, open, close, s, ans){
    if(open === n && close === n){
        ans.push(s);
        return;
    }

    if(open < n) helper(n, open+1, close, s+"(", ans);
    if(open > close) helper(n, open, close+1, s+")", ans);
}


var generateParenthesis = function(n) {
    const ans = [];
    helper(n, 0, 0, "", ans);
    return ans;
};